src/cosalib/metal.py

```diff
@@ -0,0 +1,319 @@
+VARIANTS = {
+    "qemu": {
+        "arch": "any",
+        "platform": "qemu",
+        "format": "qcow2",
+        "size": "defined",
+        "target_args": [
+            "-drive",
+            "if=virtio,id=target,format={format},file={outfile},cache=unsafe"
+        ],
+        "kargs": [
+            "console=tty0 console={terminal},115200n8",
+            "ignition.platform.id={platform}"
+        ],
+    },
+    "metal": {
+        "arch": "any",
+        "platform": "metal",
+        "format": "raw",
+        "size": "estimate",
+        "kargs": [
+            "console=tty0 console={terminal},115200n8",
+            "ignition.platform.id={platform}"
+        ],
+        "target_args": [
+            "-drive",
+            "if=virtio,id=target,format={format},file={outfile},cache=unsafe"
+        ],
+    },
+    "dasd": {
+        "arch": "s390x",
+        "format":  "raw",
+        "platform": "metal",
+        "size": "estimate",
+        "kargs": [
+            "ignition.platform.id={platform}"
+        ],
+        "target_args": [
+            "-drive",
+            "if=virtio,id=target,format={format},file={outfile},cache=unsafe"
+            "-device",
+            ("virtio-blk-ccw,drive=target,physical_block_size=4096,"
+             "logical_block_size=4096,scsi=off")
+        ],
+    }
+}
+
+ARCH_TERMINAL = {
+    "x86_64": "ttyS0",
+    "ppc64le": "hvc0",
+    "aarch64": "ttyAMA0",
+    "s390x": "ttysclp0",
+}
+
+
+def arch_terminal(arch):
+    """
+    Get the default terminal based on the architecture
+    """
+    if arch is None:
+        arch = get_basearch
+    return ARCH_TERMINAL[arch]
+
+
+class MetalImageError(Exception):
+    """
+    Base error for build issues.
+    """
+    pass
+
+
+def get_metal_variant(variant, parser, kwargs={}):
+    """
+    Helper function get get the MetalVariantImage Build Obj
+    """
+    log.debug(f"returning MetalVariantImage for {variant}")
+    return MetalVariantImage(
+        parser.buildroot,
+        parser.build,
+        variant=variant,
+        **kwargs)
+
+
+class MetalVariantImage(OSTreeBuild):
+    def __init__(self, *args, **kwargs):
+        """
+        This takes all the regular _BuildClass arguments. In kwargs, the
+        additional arguments are used:
+            platform: the name of the image.
+            format: the disk format type, e.g. raw
+            arch: build for specific architecture
+            force: force the build
+            target_args: qemu specific arguments
+
+        Alternatively, you can provide "variant=<variant>" and defaults will be
+        used. Default to building "metal" disk targets.
+        """
+        self._image_name = None
+        variant = kwargs.pop("variant", "metal")
+        if variant:
+            kwargs.update(VARIANTS.get(variant, {}))
+        self.arch = kwargs.pop("arch", get_basearch)
+        self.format = kwargs.pop("format")
+        self.platform = kwargs.pop("platform")
```

Comment:
nit: Will raise `KeyError` if not defined. May be expected (and thus OK)

Reply:
Intended :)